src/Network/Riak/Connection/Internal.hs

```diff
@@ -337,6 +346,124 @@ pipeline_ conn@Connection{..} reqs = do
   _ <- onIOException "pipeline_" $ waitBoth sendReqs receiveResps
   return ()
 
+data WorkerResult
+  = WorkerDone               -- Background request-sender has run out of requests.
+  | WorkerDied SomeException -- Background request-sender died somehow.
+
+stream
+  :: forall req resp.
+     Request req
+  => (Connection -> IO resp)
+  -> Connection
+  -> Pipes.Producer req IO ()
+  -> Pipes.Producer' (req, resp) IO ()
+stream receive conn@Connection{..} reqs = do
+  -- Keep track of the requests sent. This allows us to pair requests with their
+  -- corresponding responses.
+  requestsChan :: TChan req <-
+    lift newTChanIO
+
+  resultVar :: TMVar WorkerResult <-
+    lift newEmptyTMVarIO
+
+  -- Keep a weak reference to the worker result var: if the worker sees this var
+  -- has been garbage collected, it knows the driver of the returned producer
+  -- has died, or dropped the producer on the floor - point is, there's no one
+  -- left that cares about the responses, so don't bother sending any more
+  -- requests.
+  weakResultVar :: Weak (TMVar WorkerResult) <-
+    lift (mkWeakTMVar resultVar (pure ()))
+
+  -- Background thread: send all of the requests (or die trying).
+  let worker :: IO ()
+      worker = Pipes.runEffect (reqs >-> doSend)
+        where
+          doSend :: Pipes.Consumer' req IO ()
+          doSend = do
+            req <- Pipes.await
+            alive <- lift (isJust <$> deRefWeak weakResultVar)
+            when alive $ do
+              lift (sendRequest conn req)
```

Comment:
In this scenarios normally you want to batch requests rather than send them all at once, as in normally it will be faster to send rests at the rate that you consume rather than as fast as you can and risk overloading something. Maybe it is ok in this case, but it just stands out as an unusual pattern.

Reply:
The caller of the code may feed the `Producer` at his/her preferred rate, even using the return values streaming for backpressure if necessary.

The simplest `Producer` would indeed just be a `[request]` lifted to `Producer request IO ()`, so it's already "known" how many requests will be made (as when calling e.g. `pipeline`).

Does that help clarify anything? Or am I off base here?